sklearn/metrics/_dist_metrics.pyx.tp

```diff
@@ -749,17 +749,17 @@ cdef class DistanceMetric{{name_suffix}}(DistanceMetric):
     def _pairwise_dense_dense(self, X, Y):
         cdef const {{INPUT_DTYPE_t}}[:, ::1] Xarr
         cdef const {{INPUT_DTYPE_t}}[:, ::1] Yarr
-        cdef float64_t[:, ::1] Darr
+        cdef {{INPUT_DTYPE_t}}[:, ::1] Darr
```

Comment:
It seems that the allocations of `Darr` are the main cost, which is being reduced with the changes here. Yet, some implementations uses `float64` precision on `float32` data for computing distances; here is just one example:

https://github.com/scikit-learn/scikit-learn/blob/a16d3673fe20ef35fcb5d1f9889f14364606ed5e/sklearn/metrics/_pairwise_distances_reduction/_argkmin.pyx.tp#L278-L288

What do you think?

Have you tried only to change the allocations of `Darr` to use `INPUT_DTYPE_t` without changing the return type of the methods?

Reply:
I would also suggest to try without changing the return type.

Reply:
Sure, it doesn't mess with performance much. Quite frankly I just didn't know if the semantics made sense to me there. Accumulating in high precision makes sense, but returning a high precision type despite accumulating over low-precision data seems misleading.

I've made the changes, but I'm not convinced that these are healthy semantics.

Reply:
Actually, thinking about it more, I believe we should **not** default to `float64_t`, and instead preserve input dtype here too. We accumulate in `float64_t` to mitigate imprecision through the additional FP operations, but we do not _need_ to return `float64_t`. Doing so, imo, masks the imprecision brought on by consuming `float32_t` data in the first place. I believe this is poor practice that may end up being a disservice to users.

Users will still have the option of manually upcasting back to `float64_t` at the call-site (which is minimal overhead). This forces them to acknowledge that the computation performed is not magically more precise than the data it consumed.